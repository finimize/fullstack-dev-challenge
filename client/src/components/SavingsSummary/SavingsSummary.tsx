import React, { FC, useContext } from 'react'
import { Text } from '@chakra-ui/react'
import { AppContext } from '../../store'
import { Card } from '../Card'

export const SavingsSummary: FC = () => {
    const { state } = useContext(AppContext)

    return (
        <Card>
            <Text
                fontSize='sm'
                marginBottom='4'
                color='blueHeader'
                fontWeight='700'
                textAlign='center'
            >
                Savings Summary
            </Text>

            <Text fontSize='xl' marginBottom='4' color='grey5' fontWeight='700' textAlign='center'>
                Total Savings in 50 years
            </Text>
            <Text
                fontSize='sm'
                marginBottom='4'
                color='blueHeader'
                fontWeight='700'
                textAlign='center'
            >
                Â£ {state.data.value?.finalValue}
            </Text>
        </Card>
    )
}
