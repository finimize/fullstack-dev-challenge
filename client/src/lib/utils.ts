import type { ChartTooltipItem } from 'chart.js'
import { SAVINGS_DEFAULTS } from './constants'

export const createQueryString = (queries: { [key: string]: any }): string =>
    Object.keys(queries)
        .map((key) => encodeURIComponent(key) + '=' + encodeURIComponent(queries[key]))
        .join('&')

export const appendYearsQuery = (queryString: string) =>
    `${queryString}&years=${SAVINGS_DEFAULTS.YEARS}`

export const extractYearlyData = (projections: number[]) =>
    projections.filter((item, index) => index % 12 === 0)

export const numberFormatter = new Intl.NumberFormat('gb-GB', {
    style: 'currency',
    currency: 'GBP',
})

export const getTooltipTitle = (tooltipItems: ChartTooltipItem[]) => `Year ${tooltipItems[0].label}`

export const getTooltipLabel = (tooltipItem: ChartTooltipItem) => {
    const value = tooltipItem.value ? parseFloat(tooltipItem.value) : 0
    return numberFormatter.format(value)
}
