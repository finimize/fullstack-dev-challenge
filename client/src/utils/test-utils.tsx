import React, { FC, useReducer } from 'react'
import { ChakraProvider, extendTheme } from '@chakra-ui/react'
import { render as rtlRender, RenderResult, RenderOptions } from '@testing-library/react'
import { AppContext, DispatchType } from '../store'
import { initialState } from '../store/initialState'
import { reducer } from '../store/reducer'
import { StateInterface } from '../store/store.interface'
import { theme } from '../theme'

const mainTheme = extendTheme(theme)

interface CustomRenderOptions extends Partial<RenderOptions> {
    store?: StateInterface
    defaultTheme?: typeof theme
}

export interface CustomRenderInterface {
    (
        ui: React.ReactElement<unknown, string | React.JSXElementConstructor<unknown>>,
        options: CustomRenderOptions,
    ): { render: RenderResult; dispatchMock: jest.Mock<unknown, unknown[]> }
}

const customRender: CustomRenderInterface = (
    ui,
    { store = initialState, defaultTheme = mainTheme, ...options },
) => {
    const dispatchMock = jest.fn()
    const AllProviders: FC = ({ children }) => {
        // const [calculatorState, setCalculatorState] = useState(store)
        const [calculatorState, dispatch] = useReducer(reducer, store)
        return (
            <ChakraProvider theme={defaultTheme}>
                <AppContext.Provider
                    value={{
                        state: calculatorState,
                        dispatch: dispatchMock.mockImplementation((value: DispatchType) =>
                            dispatch(value),
                        ),
                    }}
                >
                    {' '}
                    {children}{' '}
                </AppContext.Provider>{' '}
            </ChakraProvider>
        )
    }
    return { render: rtlRender(ui, { wrapper: AllProviders, ...options }), dispatchMock }
}

export { customRender }
