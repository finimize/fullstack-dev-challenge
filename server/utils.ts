const compoundedInterest = (interestRate: number, timeInMonths: number) :number => {
    return Math.pow((interestRate + 1), timeInMonths);
};

const percentageToDecimal = (percentValue: number) :number => percentValue/100;

export const monthlyCompoundInterest = (monthlyDeposit: number, monthlyInterest: number, timeInMonths: number) :number =>  {
    const interestRate = percentageToDecimal(monthlyInterest);
    const interestOverMonths = compoundedInterest(interestRate, timeInMonths);
    const monthlyCompound = (monthlyDeposit *((interestOverMonths-1)/interestRate));
    return Math.round(monthlyCompound);
}

export const principleCompoundInterest = (initialDeposit: number,
                                          monthlyInterest: number,
                                          timeInMonths: number) : number => {
    const interestOverMonths = compoundedInterest(percentageToDecimal(monthlyInterest), timeInMonths);
    return Math.round(initialDeposit * interestOverMonths);
}

export const totalSavingsOverTime = (initialDeposit: number,
                                    monthlyDeposit: number,
                                    monthlyInterest: number,
                                    timeInMonths: number) :number => {
    const principle = principleCompoundInterest(initialDeposit, monthlyInterest, timeInMonths);
    const  monthly = monthlyCompoundInterest(monthlyDeposit, monthlyInterest, timeInMonths);
    return monthly+principle;
}

const toRange = (value: number) => {
    let array = [];
    for(let i =1;i<=value;i++){
        array.push(i);
    }
    return array;
}
export const toYearsFromNow = (years: number) :Array<string> => {
    let yearsArray = toRange(years);
    return yearsArray.map(number => {
        let d = new Date();
        let currentYear = d.getFullYear();
        let month = d.getMonth();
        let day = d.getDate();
        let c = new Date(currentYear + number, month, day);
        return c.getFullYear().toString();
    })
}
export const yearlySavingsOverTime = (initialDeposit: number,
                                        monthlyDeposit: number,
                                         monthlyInterest: number,
                                         timeInYears: number) : Array<number> => {
        const arr = toRange(timeInYears);
        const principleCompound = principleCompoundInterest(initialDeposit,monthlyInterest,timeInYears*12);
        return arr.map(number => monthlyCompoundInterest(monthlyDeposit,monthlyInterest,number*12) + principleCompound);
}
