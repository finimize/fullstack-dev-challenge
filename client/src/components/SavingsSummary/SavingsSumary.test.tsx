import React from 'react'
import { cleanup } from '@testing-library/react'
import { customRender } from '../../utils/test-utils'
import { SavingsSummary } from './SavingsSummary'

afterEach(cleanup)

describe('component/SavingsSummary', () => {
    it('renders Savings summary component', () => {
        const { render } = customRender(<SavingsSummary />, {})
        const { container } = render
        expect(container).toMatchSnapshot()
    })
})
