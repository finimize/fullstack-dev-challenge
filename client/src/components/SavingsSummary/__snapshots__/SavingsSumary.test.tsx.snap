// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/SavingsSummary renders Savings summary component 1`] = `
<div>
   
  <div
    class="css-6z87o"
  >
    <p
      class="chakra-text css-nrvl4t"
    >
      Savings Summary
    </p>
    <p
      class="chakra-text css-1hgde6q"
    >
      Total Savings in 50 years
    </p>
    <p
      class="chakra-text css-nrvl4t"
    >
      Â£ 
    </p>
  </div>
   
   
</div>
`;
